lib/AST/ASTContext.cpp

```diff
@@ -3108,8 +3108,13 @@ AnyFunctionType::Param swift::computeSelfParam(AbstractFunctionDecl *AFD,
   } else if (auto *CD = dyn_cast<ConstructorDecl>(AFD)) {
     if (isInitializingCtor) {
       // initializing constructors of value types always have an implicitly
-      // inout self.
-      isMutating = true;
+      // inout self, as do convenience initializers in Swift 5 and later.
+      if (!containerTy->hasReferenceSemantics()) {
+        isMutating = true;
+      } else if (Ctx.LangOpts.isSwiftVersionAtLeast(5)
+                 && !CD->isDesignatedInit()) {
```

Comment:
Would it be clearer to say CD->isConvenienceInit()?

Reply:
If we ever expose factory inits as a user facing feature, they’d have the same behavior, wouldn’t they?